// ngx-chronometer/esm2015/lib/ngx-chronometer.js
import { Subject } from 'rxjs';
export var TimeChonometer;
(function (TimeChonometer) {
    TimeChonometer[TimeChonometer["timemedio"] = 0] = "timemedio";
    TimeChonometer[TimeChonometer["timeone"] = 1] = "timeone";
    TimeChonometer[TimeChonometer["timetwo"] = 2] = "timetwo";
})(TimeChonometer || (TimeChonometer = {}));
export var StatusChonometer;
(function (StatusChonometer) {
    StatusChonometer[StatusChonometer["desactived"] = 0] = "desactived";
    StatusChonometer[StatusChonometer["pause"] = 1] = "pause";
    StatusChonometer[StatusChonometer["start"] = 2] = "start";
    StatusChonometer[StatusChonometer["finish"] = 3] = "finish";
    StatusChonometer[StatusChonometer["restart"] = 4] = "restart";
    StatusChonometer[StatusChonometer["stop"] = 5] = "stop";
    StatusChonometer[StatusChonometer["refresh"] = 6] = "refresh";
})(StatusChonometer || (StatusChonometer = {}));
export class Chronometer {
    /**
     * @description Constructor of class
     * @param o Object default = {}
     */
    constructor(o = {}) {
        this.time = new Array(0, 0, 0);
        this.onChronometer = new Subject();
        this.id = o.id;
        this.second = o.second || 0;
        this.status = o.status || StatusChonometer.pause;
        this.timeLabel = o.timeLabel;
        this.maxSecond = o.maxSecond || 60;
        this.maxMinute = o.maxMinute || 60;
        this.maxHour = o.maxHour || 60;
        this.limitSecond = o.limitSecond;
    }
    start() {
        this.status = StatusChonometer.start;
        this.onChronometer.next(this);
    }
    pause() {
        this.status = StatusChonometer.pause;
        this.onChronometer.next(this);
    }
    restart() {
        this.status = StatusChonometer.restart;
        this.onChronometer.next(this);
    }
    stop() {
        this.status = StatusChonometer.stop;
        this.onChronometer.next(this);
    }
    refresh() {
        this.onChronometer.next(this);
    }
}
